### PR TITLE
[Merged by Bors] - feat: port `Algebra.Order.Floor`

### PR DESCRIPTION
**Pull request by ADedecker**
This PR provided some of the first serious testing of `linarith` and `abel`, resulting in bugfixes #1358 and #1394.

- [x] depends on: #1392
- [x] depends on: #1394

### CODE REVIEW

**Comment by hrmacbeth** [comment]
The following bizarre behaviour occurs for me locally -- I wonder if anyone else can reproduce it.  As the PR stands, the `abel` call in `fract_add` fails.  If I add 
```
example (a b : α) (k l n : ℤ) : a + b + -n + (k + -a) + (l + -b) = k + l + -n := by abel
```
right before `fract_add`, then the `abel` call in `fract_add` succeeds.

**Comment by ADedecker** [comment]
Thanks a lot Heather! (I had started doing basic changes but forgot to push, so I'll just use what you did instead). Regarding the spelling for `coe`, I think the best is to use `(↑)` instead, as was suggested by Gabriel [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/How.20to.20spell.20.60coe.60.20in.20Lean4/near/318127576), but that seems to conflict with what's written in the porting wiki.
Also, the `convert` failure in `floor_pos` looks really fishy, I'll investigate that very soon.

**Comment by ADedecker** [comment]
Also, I'd say the porting notes "changed `(coe : ℕ → α)` to `(Nat.cast : ℕ → α)`" are a bit misleading, because the core problem is that `(coe : ℕ → α)` doesn't mean what it used to in Lean3. I'd argue that `Nat.cast` (or `(↑)`, these are syntactically equal) is the natural translation there, so it doesn't need a porting note. Is it okay if I remove them?

**Comment by ADedecker** [comment]
> The following bizarre behaviour occurs for me locally -- I wonder if anyone else can reproduce it. As the PR stands, the `abel` call in `fract_add` fails. If I add
> 
> ```
> example (a b : α) (k l n : ℤ) : a + b + -n + (k + -a) + (l + -b) = k + l + -n := by abel
> ```
> 
> right before `fract_add`, then the `abel` call in `fract_add` succeeds.

To add a datapoint, for me the current proof fails all the time, but the following works only with this example right above:
```lean
theorem fract_add (a b : α) : ∃ z : ℤ, fract (a + b) - fract a - fract b = z :=
  ⟨⌊a⌋ + ⌊b⌋ - ⌊a + b⌋, by
    unfold fract
    simp [sub_eq_add_neg]
    generalize (⌊a⌋ : α) = x
    abel⟩
```

**Comment by hrmacbeth** [comment]
> Regarding the spelling for `coe`, I think the best is to use `(↑)` instead, as was suggested by Gabriel [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/How.20to.20spell.20.60coe.60.20in.20Lean4/near/318127576), but that seems to conflict with what's written in the porting wiki.

Thanks for the pointer, let's do what Gabriel says (and maybe ask if we should update the porting wiki ...)

> I'd argue that `Nat.cast` is the natural translation there, so it doesn't need a porting note. Is it okay if I remove them?

Sure!

> To add a datapoint, for me the current proof fails all the time, but the following works only with this example right above:

I think that's the same behaviour as for me, then?  (Unless I have misunderstood your description.)

File: Mathlib/Algebra/Order/Floor.lean [hunk 1]
```diff
@@ -0,0 +1,1689 @@
...
+        apply add_lt_add_left
+        conv_rhs => rw [←add_halves 1]
+        norm_num
+    rw [if_neg (not_lt.mpr hx), ← fract_add_floor x, add_assoc, add_left_comm, floor_int_add,
+      ceil_add_int, add_comm _ ⌊x⌋, add_right_inj, ceil_eq_iff, this, cast_one, sub_self]
+    constructor
+    -- was linarith [fract_lt_one x]
+    · apply lt_of_lt_of_le _ hx
+      norm_num
+    · exact (fract_lt_one x).le
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:1453]: Replacing those `linarith` calls makes me sad

**Comment by hrmacbeth** [comment]
Merged in order to get the new caching mechanism.

File: Mathlib/Algebra/Order/Floor.lean (hunk 1 above)
Comment (hrmacbeth) [reply Mathlib/Algebra/Order/Floor.lean:1453]: Opened an issue for one of them:
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/linarith.20failure

**Comment by hrmacbeth** [comment]
Merged again to get the linarith fixes in #1358.

File: Mathlib/Algebra/Order/Floor.lean (hunk 1 above)
Comment (hrmacbeth) [reply Mathlib/Algebra/Order/Floor.lean]: Now that that issue has been fixed (#1358), I believe all remaining linarith failures are due to the lack of fraction functionality in `ring` (#1189).  I rewrote your fixes on the broken linarith calls in order to make this transparent.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1392
* leanprover-community/mathlib4#1394
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by LukasMias** [comment]
Linter accepts Algebra.Order.Floor now, at the cost of commenting out the ```@[simp]```s at the ```simp can prove this``` theorems. If there is a better way to fix these, feel free to revert of course.

The remaining linter problems are about docstrings in Init.Meta.WellFoundedTactics, which I don't properly understand unfortunately 😄

**Comment by hrmacbeth** [comment]
Thanks @LukasMias !  I will merge in the latest version of #1392 and that should fix Init.Meta.WellFoundedTactics.

File: Mathlib/Algebra/Order/Floor.lean [hunk 2]
```diff
@@ -0,0 +1,1704 @@
...
+  simpa only [Int.succ, Int.cast_add, Int.cast_one] using lt_succ_floor a
+#align int.lt_floor_add_one Int.lt_floor_add_one
+
+@[simp]
+theorem sub_one_lt_floor (a : α) : a - 1 < ⌊a⌋ :=
+  sub_lt_iff_lt_add.2 (lt_floor_add_one a)
+#align int.sub_one_lt_floor Int.sub_one_lt_floor
+
+@[simp]
+theorem floor_int_cast (z : ℤ) : ⌊(z : α)⌋ = z :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:701]: `intCast` / `natCast`?

File: Mathlib/Algebra/Order/Floor.lean [hunk 3]
```diff
@@ -0,0 +1,1704 @@
...
+  abs_eq_self.mpr <| fract_nonneg a
+#align int.abs_fract Int.abs_fract
+
+@[simp]
+theorem abs_one_sub_fract : |1 - fract a| = 1 - fract a :=
+  abs_eq_self.mpr <| sub_nonneg.mpr (fract_lt_one a).le
+#align int.abs_one_sub_fract Int.abs_one_sub_fract
+
+@[simp]
+theorem fract_int_cast (z : ℤ) : fract (z : α) = 0 := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:905]: Same q

File: Mathlib/Algebra/Order/Floor.lean [hunk 4]
```diff
@@ -0,0 +1,1704 @@
...
+  rw [ceil_le, Int.cast_add, Int.cast_one]
+  exact (lt_floor_add_one a).le
+#align int.ceil_le_floor_add_one Int.ceil_le_floor_add_one
+
+theorem le_ceil (a : α) : a ≤ ⌈a⌉ :=
+  gc_ceil_coe.le_u_l a
+#align int.le_ceil Int.le_ceil
+
+@[simp]
+theorem ceil_int_cast (z : ℤ) : ⌈(z : α)⌉ = z :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:1134]: And here

File: Mathlib/Algebra/Order/Floor.lean [hunk 5]
```diff
@@ -0,0 +1,1704 @@
...
+@[simp]
+theorem round_zero : round (0 : α) = 0 := by simp [round]
+#align round_zero round_zero
+
+@[simp]
+theorem round_one : round (1 : α) = 1 := by simp [round]
+#align round_one round_one
+
+@[simp]
+theorem round_nat_cast (n : ℕ) : round (n : α) = n := by simp [round]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:1351]: Here

File: Mathlib/Algebra/Order/Floor.lean [hunk 6]
```diff
@@ -0,0 +1,1704 @@
...
+
+namespace Int
+
+variable [LinearOrderedField α] [LinearOrderedField β] [FloorRing α] [FloorRing β]
+  [RingHomClass F α β] {a : α} {b : β}
+
+-- Porting note: no longer needed
+-- include β
+
+set_option linter.deprecated false in
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:1572]: Still needed?

File: Mathlib/Algebra/Order/Floor.lean [hunk 7]
```diff
@@ -0,0 +1,1704 @@
...
+section FloorRingToSemiring
+
+variable [LinearOrderedRing α] [FloorRing α]
+
+/-! #### A floor ring as a floor semiring -/
+
+
+-- see Note [lower instance priority]
+instance (priority := 100) FloorRing.toFloorSemiring : FloorSemiring α
+    where
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:1592]:
```suggestion
instance (priority := 100) FloorRing.toFloorSemiring : FloorSemiring α where
```

File: Mathlib/Algebra/Order/Floor.lean [hunk 8]
```diff
@@ -0,0 +1,1704 @@
...
+  gc_ceil a n := by rw [Int.toNat_le, Int.ceil_le, Int.cast_ofNat]
+#align floor_ring.to_floor_semiring FloorRing.toFloorSemiring
+
+theorem Int.floor_to_nat (a : α) : ⌊a⌋.toNat = ⌊a⌋₊ :=
+  rfl
+#align int.floor_to_nat Int.floor_to_nat
+
+theorem Int.ceil_to_nat (a : α) : ⌈a⌉.toNat = ⌈a⌉₊ :=
+  rfl
+#align int.ceil_to_nat Int.ceil_to_nat
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Floor.lean:1606]:
```suggestion
theorem Int.floor_toNat (a : α) : ⌊a⌋.toNat = ⌊a⌋₊ :=
  rfl
#align int.floor_to_nat Int.floor_toNat

theorem Int.ceil_toNat (a : α) : ⌈a⌉.toNat = ⌈a⌉₊ :=
  rfl
#align int.ceil_to_nat Int.ceil_toNat
```

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1392~~
* ~~leanprover-community/mathlib4#1394~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Algebra/Order/Floor.lean [hunk 9]
```diff
@@ -8,11 +8,14 @@ Authors: Mario Carneiro, Kevin Kappelmann
 ! Please do not edit these lines, except to modify the commit id
 ! if you have ported upstream changes.
 -/
+import Mathlib.Algebra.CharZero.Lemmas
 import Mathlib.Data.Int.Lemmas
+import Mathlib.Data.Int.CharZero
 import Mathlib.Data.Set.Intervals.Group
 import Mathlib.Data.Set.Lattice
+import Mathlib.Init.Meta.WellFoundedTactics
```
Comment (jcommelin) [inline Mathlib/Algebra/Order/Floor.lean:16]: Why do we need these new imports, compared to mathlib3?

Comment (hrmacbeth) [reply Mathlib/Algebra/Order/Floor.lean:16]: I believe they were formerly imported indirectly through `Tactic.Positivity` or `Tactic.Linarith`.  Note that the mathlib4 versions of those tactics are currently lacking a lot of functionality compared to their mathlib3 versions, so probably they import less.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**